tree/ntuple/v7/inc/ROOT/RMiniFile.hxx

```diff
@@ -0,0 +1,221 @@
+   RNTuple GetNTuple(std::string_view ntupleName);
+   /// Reads a given byte range from the file into the provided memory buffer
+   void ReadBlob(void *buffer, size_t nbytes, std::uint64_t offset);
+};
+
+
+// clang-format off
+/**
+\class ROOT::Experimental::Internal::RMiniFileWriter
+\ingroup NTuple
+\brief Write RNTuple data blocks in a TFile container
+
+The writer can create a new TFile container for an RNTuple or add an RNTuple to an existing TFile.
+Creating a single RNTuple in a new TFile container can be done with a C file stream without a TFile class.
+Updaing an existing TFile requires a proper TFile object.  Also, writing a remote file requires a proper TFile object.
+A stand-alone version of RNTuple can remove the TFile based writer.
+*/
+// clang-format on
+class RMiniFileWriter {
+private:
+   struct RFileProper {
+      TFile *fFile = nullptr;
+      /// Low-level writing using a TFile
+      void Write(const void *buffer, size_t nbytes, std::int64_t offset);
+      /// Writes an RBlob opaque key with the provided buffer as data record and returns the offset of the record
+      std::uint64_t WriteKey(const void *buffer, size_t nbytes, size_t len);
+      operator bool() const { return fFile; }
+   };
+
+   struct RFileSimple {
+      /// For the simplest cases, a C file stream can be used for writing
+      FILE *fFile = nullptr;
+      /// Keeps track of the seek offset
+      std::uint64_t fFilePos = 0;
+
+      RFileSimple() = default;
+      RFileSimple(const RFileSimple &other) = delete;
+      RFileSimple(RFileSimple &&other) = delete;
+      RFileSimple &operator =(const RFileSimple &other) = delete;
+      RFileSimple &operator =(RFileSimple &&other) = delete;
+      ~RFileSimple();
+
+      /// Writes bytes in the open stream, either at fFilePos or at the given offset
+      void Write(const void *buffer, size_t nbytes, std::int64_t offset = -1);
+      /// Writes a TKey including the data record, given by buffer, into fFile; returns the file offset to the payload.
+      /// The payload is already compressed
+      std::uint64_t WriteKey(const void *buffer, std::size_t nbytes, std::size_t len, std::int64_t offset = -1,
+                             std::uint64_t directoryOffset = 100,
+                             const std::string &className = "",
+                             const std::string &objectName = "",
+                             const std::string &title = "");
+      operator bool() const { return fFile; }
+   };
+
+   // TODO(jblomer): wrap in an std::variant with C++17
+   /// For updating existing files and for storing more than just an RNTuple in the file
+   RFileProper fFileProper;
+   /// For simple use cases, survives without libRIO dependency
+   RFileSimple fFileSimple;
+   /// A simple file can either be written as TFile container or as NTuple raw file
+   bool fIsRaw = false;
+   /// The identifier of the RNTuple
+   std::string fNTupleName;
```

Comment:
This implies that there is only one RNTuple per TFile.    Is that a temporary restriction?

Reply:
I added a clarification in the member comment

Reply:
> writers can concurrently operate on the same file iff they use a proper TFile object for writing.

And the proper 'locks' (in case of multiple thread) as neither File nor TFile are thread safe.

> iff they use a proper

Do you mean "if"?

Reply:
The "concurrently" is misleading, I removed it.  What I mean is that multiple `RNTupleWriter` objects can co-exist in a sequential program, in which case they'd need to share the `TFile` object of course.

Replaced "iff" by "if (and only if)".